File: log-view-machine/java/src/test/java/com/logview/causality/UndoRedoIntegrationTest.java
package com.logview.causality;

import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class UndoRedoIntegrationTest {

    @Test
    void idleToCookingToCompletedThenUndoTwiceRedoOnce() {
        MachineConfig<Integer> config = new MachineConfig<>(
            "idle",
            0,
            Map.of(
                "idle", new StateNodeConfig(Map.of("START", TransitionTarget.state("cooking"))),
                "cooking", new StateNodeConfig(Map.of("DONE", TransitionTarget.state("completed"))),
                "completed", new StateNodeConfig(Map.of("RESET", TransitionTarget.state("idle")))
            )
        );
        Interpreter<Integer> service = Interpreter.interpret(StateMachine.create(config));

        service.send("START");
        assertEquals("cooking", service.getValue());
        service.send("DONE");
        assertEquals("completed", service.getValue());

        assertTrue(service.undo());
        assertEquals("cooking", service.getValue());
        assertTrue(service.undo());
        assertEquals("idle", service.getValue());

        assertTrue(service.redo());
        assertEquals("cooking", service.getValue());
        assertTrue(service.canRedo()); // one more redo (to completed) available
        assertTrue(service.redo());
        assertEquals("completed", service.getValue());
        assertFalse(service.canRedo());
    }

    @Test
    void sendAfterUndoClearsRedo() {
        MachineConfig<Integer> config = new MachineConfig<>(
            "idle",
            0,
            Map.of(
                "idle", new StateNodeConfig(Map.of("GO", TransitionTarget.state("active"))),
                "active", new StateNodeConfig(Map.of("BACK", TransitionTarget.state("idle")))
            )
        );
        Interpreter<Integer> service = Interpreter.interpret(StateMachine.create(config));

        service.send("GO");
        assertEquals("active", service.getValue());
        service.undo();
        assertEquals("idle", service.getValue());
        assertTrue(service.canRedo());

        service.send("GO");
        assertFalse(service.canRedo());
        assertEquals("active", service.getValue());
    }

    @Test
    void setStateOverrideWritesNodeWithOverrideEventAndUndoToRootReturnsToInitial() {
        MachineConfig<Integer> config = new MachineConfig<>(
            "idle",
            0,
            Map.of(
                "idle", new StateNodeConfig(Map.of("START", TransitionTarget.state("cooking"))),
                "cooking", new StateNodeConfig(Map.of("DONE", TransitionTarget.state("completed"))),
                "completed", new StateNodeConfig(Map.of("RESET", TransitionTarget.state("idle")))
            )
        );
        Interpreter<Integer> service = Interpreter.interpret(StateMachine.create(config));
        assertEquals("idle", service.getValue());

        service.setState("completed", 99);
        StateSnapshot<Integer> snap = service.getSnapshot();
        assertEquals("completed", snap.getValue());
        assertEquals(99, snap.getContext());
        assertEquals(MachineConfig.OVERRIDE_EVENT, snap.getEvent());

        service.undoToRoot();
        assertEquals("idle", service.getValue());
        assertEquals(0, service.getContext());
    }

    @Test
    void backwardIsAliasForUndoToRoot() {
        MachineConfig<Integer> config = new MachineConfig<>(
            "idle",
            0,
            Map.of(
                "idle", new StateNodeConfig(Map.of("GO", TransitionTarget.state("active")))
            )
        );
        Interpreter<Integer> service = Interpreter.interpret(StateMachine.create(config));
        service.send("GO");
        assertEquals("active", service.getValue());
        assertTrue(service.backward());
        assertEquals("idle", service.getValue());
        assertFalse(service.backward());
    }
}

File: log-view-machine/core/src/test/kotlin/com/logview/causality/UndoRedoIntegrationTest.kt
package com.logview.causality

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*

class UndoRedoIntegrationTest {

    @Test
    fun `idle to cooking to completed then undo twice redo once`() {
        val config = MachineConfig(
            initial = "idle",
            context = 0,
            states = mapOf(
                "idle" to StateNodeConfig(on = mapOf("START" to TransitionTarget.state("cooking"))),
                "cooking" to StateNodeConfig(on = mapOf("DONE" to TransitionTarget.state("completed"))),
                "completed" to StateNodeConfig(on = mapOf("RESET" to TransitionTarget.state("idle")))
            )
        )
        val service = interpret(createMachine(config))

        service.send("START")
        assertEquals("cooking", service.value)
        service.send("DONE")
        assertEquals("completed", service.value)

        assertTrue(service.undo())
        assertEquals("cooking", service.value)
        assertTrue(service.undo())
        assertEquals("idle", service.value)

        assertTrue(service.redo())
        assertEquals("cooking", service.value)
        assertFalse(service.canRedo())
    }

    @Test
    fun `send after undo clears redo`() {
        val config = MachineConfig(
            initial = "idle",
            context = 0,
            states = mapOf(
                "idle" to StateNodeConfig(on = mapOf("GO" to TransitionTarget.state("active"))),
                "active" to StateNodeConfig(on = mapOf("BACK" to TransitionTarget.state("idle")))
            )
        )
        val service = interpret(createMachine(config))

        service.send("GO")
        assertEquals("active", service.value)
        service.undo()
        assertEquals("idle", service.value)
        assertTrue(service.canRedo())

        service.send("GO")
        assertFalse(service.canRedo())
        assertEquals("active", service.value)
    }

    @Test
    fun `setState override writes node with OVERRIDE event and undoToRoot returns to initial`() {
        val config = MachineConfig(
            initial = "idle",
            context = 0,
            states = mapOf(
                "idle" to StateNodeConfig(on = mapOf("START" to TransitionTarget.state("cooking"))),
                "cooking" to StateNodeConfig(on = mapOf("DONE" to TransitionTarget.state("completed"))),
                "completed" to StateNodeConfig(on = mapOf("RESET" to TransitionTarget.state("idle")))
            )
        )
        val service = interpret(createMachine(config))
        assertEquals("idle", service.value)

        service.setState("completed", 99)
        val snap = service.getSnapshot()
        assertEquals("completed", snap.value)
        assertEquals(99, snap.context)
        assertEquals(OVERRIDE_EVENT, snap.event)

        service.undoToRoot()
        assertEquals("idle", service.value)
        assertEquals(0, service.context)
    }

    @Test
    fun `backward is alias for undoToRoot`() {
        val config = MachineConfig(
            initial = "idle",
            context = 0,
            states = mapOf(
                "idle" to StateNodeConfig(on = mapOf("GO" to TransitionTarget.state("active")))
            )
        )
        val service = interpret(createMachine(config))
        service.send("GO")
        assertEquals("active", service.value)
        assertTrue(service.backward())
        assertEquals("idle", service.value)
        assertFalse(service.backward())
    }

    @Test
    fun `TriLeaf forward mutates head and backward clears to root`() {
        val config = MachineConfig(
            initial = "idle",
            context = 0,
            states = mapOf(
                "idle" to StateNodeConfig(on = mapOf("START" to TransitionTarget.state("cooking"))),
                "cooking" to StateNodeConfig(on = mapOf("DONE" to TransitionTarget.state("completed")))
            )
        )
        val tri = TriLeafInterpreter.interpret(createMachine(config))
        assertTrue(tri.isAtRoot())
        assertTrue(tri.getEnabledEvents().contains("START"))
        assertTrue(tri.getEnabledEvents().contains(OVERRIDE_EVENT))

        tri.forward("START")
        assertEquals("cooking", tri.value)
        assertFalse(tri.isAtRoot())
        assertTrue(tri.canBackward())

        tri.forward("DONE")
        assertEquals("completed", tri.value)
        assertTrue(tri.backward("DONE"))
        assertEquals("idle", tri.value)
        assertTrue(tri.isAtRoot())
    }

    @Test
    fun `TriLeaf forwardOverride writes override event`() {
        val config = MachineConfig(
            initial = "idle",
            context = 0,
            states = mapOf(
                "idle" to StateNodeConfig(on = mapOf("START" to TransitionTarget.state("cooking")))
            )
        )
        val tri = TriLeafInterpreter.interpret(createMachine(config))
        tri.forwardOverride("completed", 42)
        assertEquals("completed", tri.value)
        assertEquals(42, tri.context)
        assertEquals(OVERRIDE_EVENT, tri.getSnapshot().event)
        assertTrue(tri.backward(OVERRIDE_EVENT))
        assertEquals("idle", tri.value)
    }
}
